Implement sampling from the unit sphere and circle

vks opened the pull request:
This uses a method by Marsaglia (1972), which was found to be the fastest of the methods given by [MathWorld]. (See [my benchmarks](https://github.com/vks/sample-sphere).)

[MathWorld]: http://mathworld.wolfram.com/SpherePointPicking.html

There are some open questions:
* Technically this is a uniform distribution. However, for more than one dimension, the uniform unit distribution is not unique, so I implemented it as a separate distribution.
* The name `UnitSphere` is slightly incorrect. We are not sampling from the the unit sphere, but rather its surface. However, `UnitSphereSurface` is a bit long and sampling from the unit sphere is trivial given a sample from the surface, so I'm not sure we should add it.
* I did not bother to port the benchmarks to Rand. I think it is not really necessary, but it could be easily done.
* The tests don't check for uniformity.
* `assert_almost_eq` could probably be used in all tests.

dhardy commented on src/distributions/unit_sphere.rs:
```diff
@@ -0,0 +1,81 @@
+                continue;
+            }
+            let factor = (1.0_f64 - x1_sq - x2_sq).sqrt();
+            return [2.*x1 * factor, 2.*x2 * factor, 1. - 2.*sum];
```
pre-multiply factor by 2?

dhardy commented on src/distributions/mod.rs:
```diff
@@ -190,6 +193,7 @@ pub use self::bernoulli::Bernoulli;
 pub mod uniform;
 mod bernoulli;
 #[cfg(feature="alloc")] mod weighted;
+#[cfg(feature="std")]mod unit_sphere;
```
missing a space

vks commented:
Thanks for the feedback, I addressed your comments.

Do we want to add sampling from the unit sphere volume as well? The implementation is fairly straightforward (not very surprisingly, rejection sampling is the most efficient). If yes, I think `UnitSphere` should be renamed to `UnitSphereSurface`.

dhardy commented:
To be honest I'm out of my depth regarding which sampling algorithms we should choose to implement. But I read the paper for this one and it looks like a nice approach to an *apparently* common problem.

I don't *think* there's much use for sampling from the volume of a sphere ([this](https://stackoverflow.com/questions/21240768/calculating-volume-of-n-ball-using-monte-carlo-method) is about the only thing I can think of), and it's not exactly difficult to implement rejection sampling, so I don't think we would want it.

One can try to generalise sampling from the surface to other dimensions, but for 2D there's a choice of `(cos(θ), sin(θ))` or `(x, sign*sqrt(1-x^2))` (where `sign` is extracted from a spare bit), and Marsaglia's approach in 4D (from the same paper) is significantly different. Besides which, the 2D and 3D algorithms are applicable to different problems and I'm not sure if there are any common uses for 4D sampling (representing random spin using Quarternions, maybe? Too long since I looked at that problem).

So the only important questions are, I think,
- do we also want to include `UnitCircle` and `UnitHypersphere`?
- future of distributions #290 (which does not need to be addressed now)

vks commented:
> To be honest I'm out of my depth regarding which sampling algorithms we should choose to implement.

I [implemented all the common ones](https://github.com/vks/sample-sphere) and picked the fastest one. I think speed is the only concern, they should all produce correct uniform distributions.

> do we also want to include UnitCircle

FWIW, I already implemented it. I think we should include it, because the efficient implementation is non-trivial.

> and UnitHypersphere

Sampling from the volume or the surface? I can't spontaneously think of any use cases. For large dimensions, rejection sampling becomes inefficient, so it might be useful to provide an implementation.

dhardy commented:
> because the efficient implementation is non-trivial.

Only slightly surprising. Trig is slow. Do you want to try the method I suggested? Problem is getting the sign bit without an extra RNG call requires re-implementing the conversion to FP (as in our Ziggurat function).

vks commented:
> Do you want to try the method I suggested?

It is fast, but it is unfortunately not uniform.

> Problem is getting the sign bit without an extra RNG call requires re-implementing the conversion to FP (as in our Ziggurat function).

Maybe we should implement `Distribution<(f64, u8)>` for the float distributions.

vks commented:
I have a uniformity test using histograms. Should I use an external crate for the histogram implementation, or should we ship our own (private) implementation in rand?

vks commented:
> Replacing x by sin(x) should fix it, but will probably be quite slow.

Yes, that works. It is twice as fast as the naive trigonometric approach, but still slower than the current implementation.

> Not really sure what you mean here. An extra lib for testing using histograms?

No, I was using an [extra lib](https://docs.rs/average/0.9.1/average/#calculating-histograms) just for histograms, because I already implemented them there. In principle I could copy-and-paste the implementation into Rand's test directory.

vks commented on src/distributions/unit_circle.rs:
```diff
@@ -0,0 +1,85 @@
+use Rng;
+use distributions::{Distribution, Uniform};
+
+/// Samples uniformly from the surface of the unit circle in two dimensions.
```
That was a copy-and-paste leftover, sorry. I fixed it.

clarcharr commented:
Is there a reason why this is always `f64` and not parameterised to accept both `f32` and `f64`? That way in the future an implementation for say `f80` or `f128` could be added in a not breaking way.

dhardy commented:
Perhaps I should have been clearer: we're sampling from the *edge* of the unit circle. (Since we return a pair of coordinates, it would be easy to mis-understand this as sampling from the disc.)

In this case, it also seems there are [people](https://stackoverflow.com/questions/5837572/generate-a-random-point-within-a-circle-uniformly) who want to sample from the disc itself, though I don't know if this has an application or is just a toy problem. But if we ever want to implement that, we can use the name `UnitDisc` (or Disk; I'm not actually sure if that's the correct US spelling or mostly applicable to bit-storage devices; Disc is however correct *English*).

@vks can you squash please? I don't think this needs 6 commits.

> Is there a reason why this is always f64

Status quo. See #100. I'd rather get this right once and break stuff as necessary than use a piecemeal solution.

dhardy commented:
Oh, and CI is fixed now... but is tripping up on `average`. Could you please fix that too @vks?

vks commented:
> In this case, it also seems there are people who want to sample from the disc itself, though I don't know if this has an application or is just a toy problem.

[The Unity engine offers sampling from the inside of the unit circle and sphere](https://docs.unity3d.com/ScriptReference/Random.html), so I guess it is used in 3D games.

> Perhaps I should have been clearer: we're sampling from the edge of the unit circle. (Since we return a pair of coordinates, it would be easy to mis-understand this as sampling from the disc.)

The language that I was using is the following:
* "circle" = { (x, y) | x^2 + y^2 = R }
* "disk" = { (x, y) | x^2 + y^2 <= R }
* "sphere surface" = { (x, y, z) | x^2 + y^2 + z^2 = R }
* "sphere volume" = { (x, y, z) | x^2 + y^2 + z^2 <= R }

I thought that would be fairly clear, but since we don't have `UnitDisk`, maybe I should be more explicit?

I'll look into the remaining failure and rebasing/squashing again.

vks commented:
I clarified we sample from the edge of the circle, fixed the type inference failure and rebased/squashed.

dhardy commented:
Great.

We could rename `UnitSphereSurface` to `UnitShell`, but this is not very descriptive of the shape, so probably not the best idea.
